### PR TITLE
FEATURE: LX-483 Introduce Visualization Switcher widget component

### PR DESCRIPTION
**Pull request by NestorEncinas**
risk: low
JIRA: LX-483

---

> [!IMPORTANT]
> Please, **don't forget to run `rush change`** for the commits that introduce **new features** 🙏

---

Refer to [documentation](https://github.com/gooddata/gooddata-ui-sdk/blob/master/dev_docs/continuous_integration.md) to see how to run checks and tests in the pull request. This is the list of the most used commands:

```
extended test - backstop
```

```
extended test - tiger-cypress - integrated
extended test - tiger-cypress - isolated
extended test - tiger-cypress - record
```

### CODE REVIEW

**Comment by NestorEncinas** [comment]
extended test - backstop

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/2706/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by NestorEncinas** [comment]
extended test - backstop

**Comment by NestorEncinas** [comment]
extended test - backstop

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/2707/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

File: libs/sdk-ui-dashboard/src/presentation/dashboard/types.ts [hunk 1]
```diff
@@ -220,6 +221,19 @@ export interface IDashboardCustomComponentProps {
      */
     RichTextComponentProvider?: OptionalRichTextComponentProvider;
 
+    /**
+     * Specify function to obtain custom component to use for rendering a visualization switcher.
+     *
+     * @remarks
+     * -  If not provided, the default implementation {@link DefaultDashboardVisualizationSwitcher} will be used.
+     * -  If factory function is provided and it returns undefined, then the default implementation {@link DefaultDashboardVisualizationSwitcher}.
+     *    This is useful if you want to customize just one particular rich text and keep default rendering for
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-dashboard/src/presentation/dashboard/types.ts:230]: remove rich text mentions in comment

File: libs/sdk-ui-dashboard/src/presentation/dashboardContexts/DashboardComponentsContext.tsx [hunk 2]
```diff
@@ -107,6 +111,7 @@ const DashboardComponentsContext = createContext<IDashboardComponentsContext>({
     InsightMenuTitleComponentProvider: ThrowMissingComponentError("InsightMenuTitleComponent"),
     KpiComponentProvider: ThrowMissingComponentError("KpiComponent"),
     RichTextComponentProvider: ThrowMissingComponentError("RichTextComponent"),
+    VisualizationSwitcherComponentProvider: ThrowMissingComponentError("VisualizationSwticherComponent"),
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-dashboard/src/presentation/dashboardContexts/DashboardComponentsContext.tsx:114]: type Swticher -> Switcher

File: libs/sdk-ui-dashboard/src/presentation/dashboardContexts/types.ts [hunk 3]
```diff
@@ -144,6 +146,20 @@ export type RichTextComponentProvider = (widget: IRichTextWidget) => CustomDashb
  */
 export type OptionalRichTextComponentProvider = OptionalProvider<RichTextComponentProvider>;
 
+// NESTOR
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-dashboard/src/presentation/dashboardContexts/types.ts:149]: forgotten comment

File: libs/sdk-ui-dashboard/src/presentation/dragAndDrop/draggableWidget/useNewSectionVisualizationSwitcherPlaceholderDropHandler.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        addNewSectionWithVisualizationSwitcherPlaceholder(sectionIndex, {}, [
+            {
+                type: "IDashboardLayoutItem",
+                size: {
+                    xl: {
+                        gridHeight: VISUALIZATION_SWITCHER_WIDGET_SIZE_INFO_DEFAULT.height.default!,
+                        gridWidth: VISUALIZATION_SWITCHER_WIDGET_SIZE_INFO_DEFAULT.width.default!,
+                    },
+                },
+                // NESTOR
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-dashboard/src/presentation/dragAndDrop/draggableWidget/useNewSectionVisualizationSwitcherPlaceholderDropHandler.ts:40]: Forgotten comment

File: libs/sdk-ui-dashboard/src/presentation/widget/visualizationSwitcher/DefaultDashboardVisualizationSwitcher.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+// (C) 2024 GoodData Corporation
+
+import { renderModeAware } from "../../componentDefinition/index.js";
+import { EditModeDashboardVisualizationSwitcher } from "./EditModeDashboardVisualizationSwitcher.js";
+import { ViewModeDashboardVisualizationSwitcher } from "./ViewModeDashboardVisualizationSwitcher.js";
+
+/**
+ * Default implementation of the dashboard rich text widget.
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-dashboard/src/presentation/widget/visualizationSwitcher/DefaultDashboardVisualizationSwitcher.ts:8]: remove rich text mention

File: libs/sdk-ui-dashboard/src/presentation/widget/visualizationSwitcher/types.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    /**
+     * Definition of insight widget to render.
+     *
+     * @public
+     */
+    widget: IVisualizationSwitcherWidget;
+
+    /**
+     * Height of the rich text widget container.
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-dashboard/src/presentation/widget/visualizationSwitcher/types.ts:47]: remove rich text mentions

**Review by ivan-nejezchleb — CHANGES_REQUESTED** [review]

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/2708/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

File: libs/sdk-model/api/sdk-model.api.md [hunk 7]
```diff
@@ -3687,6 +3694,22 @@ export interface IVisualizationClassBody {
     url: string;
 }
 
+// @public (undocumented)
+export interface IVisualizationSwitcherWidget extends IVisualizationSwitcherWidgetBase, IDashboardObjectIdentity {
+}
+
+// @public (undocumented)
+export interface IVisualizationSwitcherWidgetBase extends IAnalyticalWidget {
+    readonly selectedVisualization: string;
```
[the remark this answers is not in the document]
Comment (NestorEncinas) [reply libs/sdk-model/api/sdk-model.api.md]: I'd like to keep as is, it doesn't "harm" besides the initial setup + avoids having to extend widgets factory.

File: libs/sdk-ui-dashboard/src/model/store/config/configSelectors.ts [hunk 8]
```diff
@@ -734,3 +734,15 @@ export const selectEnableDrillDownIntersectionIgnoredAttributes: DashboardSelect
...
+
+/**
+ * Returns whether rich text widgets are enabled.
+ *
+ * @internal
+ */
+export const selectEnableVisualizationSwitcher: DashboardSelector<boolean> = createSelector(
+    selectConfig,
+    (state) => {
+        return state.settings?.enableKDVisualizationSwitcher ?? true;
```
[the remark this answers is not in the document]
Comment (NestorEncinas) [reply libs/sdk-ui-dashboard/src/model/store/config/configSelectors.ts]: Changed to false.

File: libs/sdk-ui-dashboard/src/presentation/dashboard/types.ts (hunk 1 above)
Comment (NestorEncinas) [reply libs/sdk-ui-dashboard/src/presentation/dashboard/types.ts]: Removed.

File: libs/sdk-model/api/sdk-model.api.md [hunk 9]
```diff
@@ -3687,6 +3694,22 @@ export interface IVisualizationClassBody {
     url: string;
 }
 
+// @public (undocumented)
+export interface IVisualizationSwitcherWidget extends IVisualizationSwitcherWidgetBase, IDashboardObjectIdentity {
+}
+
+// @public (undocumented)
+export interface IVisualizationSwitcherWidgetBase extends IAnalyticalWidget {
```
[the remark this answers is not in the document]
Comment (NestorEncinas) [reply libs/sdk-model/api/sdk-model.api.md:3702]: I'd like to keep it as is, it does not "harm" much besides initial setup + will avoid the extend of widget factory.

File: libs/sdk-ui-dashboard/src/presentation/dashboardContexts/DashboardComponentsContext.tsx (hunk 2 above)
Comment (NestorEncinas) [reply libs/sdk-ui-dashboard/src/presentation/dashboardContexts/DashboardComponentsContext.tsx]: FIxed.

File: libs/sdk-ui-dashboard/src/presentation/dashboardContexts/types.ts (hunk 3 above)
Comment (NestorEncinas) [reply libs/sdk-ui-dashboard/src/presentation/dashboardContexts/types.ts]: Removed.

File: libs/sdk-ui-dashboard/src/presentation/dragAndDrop/draggableWidget/useNewSectionVisualizationSwitcherPlaceholderDropHandler.ts (hunk 4 above)
Comment (NestorEncinas) [reply libs/sdk-ui-dashboard/src/presentation/dragAndDrop/draggableWidget/useNewSectionVisualizationSwitcherPlaceholderDropHandler.ts]: Removed.

File: libs/sdk-ui-dashboard/src/presentation/widget/visualizationSwitcher/DefaultDashboardVisualizationSwitcher.ts (hunk 5 above)
Comment (NestorEncinas) [reply libs/sdk-ui-dashboard/src/presentation/widget/visualizationSwitcher/DefaultDashboardVisualizationSwitcher.ts]: Fixed,

File: libs/sdk-ui-dashboard/src/presentation/widget/visualizationSwitcher/types.ts (hunk 6 above)
Comment (NestorEncinas) [reply libs/sdk-ui-dashboard/src/presentation/widget/visualizationSwitcher/types.ts]: Fixed.

**Review by ivan-nejezchleb — APPROVED** [review]

**Pull request merged**